src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp

```diff
@@ -1047,6 +1048,127 @@ class LogicalExpressionOptimizerVisitor : public InDepthQueryTreeVisitorWithCont
         function_node.resolveAsFunction(and_function_resolver);
     }
 
+    void tryOptimizeAndCompareChain(QueryTreeNodePtr & node)
+    {
+        if (node->getNodeType() != QueryTreeNodeType::FUNCTION)
+            return;
+
+        auto & function_node = node->as<FunctionNode &>();
+        if (function_node.getFunctionName() != "and" || function_node.getResultType()->isNullable())
+            return;
+
+        enum CompareType
+        {
+            less = 0,
+            lessOrEquals,
+            equals
+        };
+
+        /// Step 1: identify constants, and store comparing pairs in hash
+        QueryTreeNodes constants;
+        /// Record a>b, a>=b, a==b pairs
+        using QueryTreeNodeWithEquals = std::vector<std::pair<QueryTreeNodePtr, CompareType>>;
+        std::unordered_map<QueryTreeNodePtr, QueryTreeNodeWithEquals> greater_pairs;
```

Comment:
looks like we only consider greater , but what about lesser?

Reply:
and if we generate many condition like `a < 5` and `a < 3` and `a < 0`, will we only keep the smallest one `a < 0`?  I think we can make it as another optimization.

Reply:
Actually we support `<, <=, >, >=, =`. For `<` and `<=` all transform to `>` and `>=`, the same to `=`. So this is why I name it `greater_pairs`, maybe suggest better name?

I tried to check early conflicts such as `a>b and a<b` when solving the loop problem of `(a > b) AND (b > c) AND (c > a) AND (a < 5)`, so that the loop won't occur at all. But as we support `>=` and `=` now, I didn't find an algorithm to detect them.

But we definitely can add another optimization to check conflict containing only `<` and `>`. Similarly, the comparing bloating issue can also be optimized before the the current optimization or after.

Reply:
Good catch, I misunderstood at first. Now it supports `a > b > c > 5`. For the conflicts, has been already fixed. It is actually not crash, instead, it goes into an endless loop.